rust-runtime/aws-smithy-runtime-api/src/client/retries.rs

```diff
@@ -3,32 +3,79 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+use crate::client::interceptors::context::Error;
 use crate::client::interceptors::InterceptorContext;
-use crate::client::orchestrator::{BoxError, RetryStrategy};
+use crate::client::orchestrator::{BoxError, HttpRequest, HttpResponse};
 use crate::config_bag::ConfigBag;
-use aws_smithy_http::body::SdkBody;
+use aws_smithy_types::retry::ErrorKind;
+use std::fmt::Debug;
+use std::time::Duration;
 
-pub mod rate_limiting;
+/// An answer to the question "should I make a request attempt?"
+pub enum ShouldAttempt {
+    Yes,
+    No,
+    YesAfterDelay(Duration),
+}
+
+pub trait RetryStrategy: Send + Sync + Debug {
+    fn should_attempt_initial_request(&self, cfg: &ConfigBag) -> Result<ShouldAttempt, BoxError>;
+
+    fn should_attempt_retry(
+        &self,
+        context: &InterceptorContext<HttpRequest, HttpResponse>,
+        cfg: &ConfigBag,
+    ) -> Result<ShouldAttempt, BoxError>;
+}
+
+#[non_exhaustive]
+#[derive(Eq, PartialEq, Debug)]
+pub enum RetryReason {
+    Error(ErrorKind),
+    Explicit(Duration),
+}
+
+/// Classifies what kind of retry is needed for a given [`Error`].
+pub trait ClassifyRetry: Send + Sync + Debug {
+    /// Run this classifier against an error to determine if it should be retried. Returns
+    /// `Some(RetryKind)` if the error should be retried; Otherwise returns `None`.
+    fn classify_retry(&self, error: &Error) -> Option<RetryReason>;
+}
 
-#[derive(Debug, Clone)]
-pub struct NeverRetryStrategy {}
+#[derive(Debug)]
+pub struct RetryClassifiers {
+    inner: Vec<Box<dyn ClassifyRetry>>,
+}
 
-impl NeverRetryStrategy {
+impl RetryClassifiers {
     pub fn new() -> Self {
-        Self {}
+        Self {
+            inner: Vec::with_capacity(1),
```

Comment:
Is it worth adding a comment for why the capacity is 1 (unless it's obvious)?

Reply:
It's always expected that at least one classifier will be defined, so we eagerly allocate for it.

I'll add a comment for this.